File: create_admin.py
#!/usr/bin/env python3
"""
Script สำหรับสร้าง Admin User ใหม่
Usage: python create_admin.py
"""

import asyncio
import sys
import os
from getpass import getpass

# เพิ่ม backend directory ให้ Python path
sys.path.append(os.path.join(os.path.dirname(__file__), 'backend'))

from lib.mongodb import connect_to_mongo, close_mongo_connection
from services.user_service import user_service
from models.user import UserRole

async def create_admin_interactive():
    """สร้าง admin user แบบ interactive"""
    print("=== MFU Learn AI - Admin User Creation ===")
    print()
    
    # รับข้อมูลจาก user
    username = input("Username: ").strip()
    if not username:
        print("❌ Username จำเป็นต้องใส่")
        return False
    
    # ตรวจสอบว่า username มีอยู่แล้วหรือไม่
    existing_user = await user_service.find_admin_by_username(username)
    if existing_user:
        print(f"❌ Username '{username}' มีอยู่แล้ว")
        return False
    
    password = getpass("Password: ").strip()
    if not password:
        print("❌ Password จำเป็นต้องใส่")
        return False
    
    confirm_password = getpass("Confirm Password: ").strip()
    if password != confirm_password:
        print("❌ Password ไม่ตรงกัน")
        return False
    
    email = input("Email: ").strip()
    if not email:
        print("❌ Email จำเป็นต้องใส่")
        return False
    
    first_name = input("First Name: ").strip()
    if not first_name:
        print("❌ First Name จำเป็นต้องใส่")
        return False
    
    last_name = input("Last Name: ").strip()
    if not last_name:
        print("❌ Last Name จำเป็นต้องใส่")
        return False
    
    department = input("Department: ").strip()
    if not department:
        print("❌ Department จำเป็นต้องใส่")
        return False
    
    # แสดงข้อมูลที่จะสร้าง
    print("\n=== ข้อมูล Admin ที่จะสร้าง ===")
    print(f"Username: {username}")
    print(f"Email: {email}")
    print(f"Name: {first_name} {last_name}")
    print(f"Department: {department}")
    print(f"Role: {UserRole.ADMIN}")
    print()
    
    confirm = input("ยืนยันการสร้าง Admin? (y/N): ").strip().lower()
    if confirm not in ['y', 'yes']:
        print("❌ ยกเลิกการสร้าง Admin")
        return False
    
    try:
        # สร้าง admin user โดยตรงโดยไม่ผ่าน create_admin method
        from lib.mongodb import get_database
        from utils.security import get_password_hash
        from datetime import datetime
        from bson import ObjectId
        
        db = get_database()
        
        # ตรวจสอบ username ซ้ำอีกครั้ง
        if await db.get_collection("users").find_one({"username": username}):
            print(f"❌ Username '{username}' มีอยู่แล้ว")
            return False
        
        # สร้าง admin document
        hashed_password = get_password_hash(password)
        admin_doc = {
            "_id": ObjectId(),
            "nameID": username,
            "username": username,
            "password": hashed_password,
            "email": email,
            "firstName": first_name,
            "lastName": last_name,
            "department": department,
            "role": UserRole.ADMIN,
            "groups": ['Admin'],
            "created": datetime.utcnow(),
            "updated": datetime.utcnow()
        }
        
        # Insert เข้าฐานข้อมูล
        await db.get_collection("users").insert_one(admin_doc)
        
        print("\n✅ สร้าง Admin สำเร็จ!")
        print(f"ID: {str(admin_doc['_id'])}")
        print(f"Username: {admin_doc['username']}")
        print(f"Email: {admin_doc['email']}")
        print(f"Name: {admin_doc['firstName']} {admin_doc['lastName']}")
        print(f"Department: {admin_doc['department']}")
        print(f"Role: {admin_doc['role']}")
        print(f"Created: {admin_doc['created']}")
        
        return True
        
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาด: {str(e)}")
        return False

async def create_super_admin():
    """สร้าง Super Admin แรก (ใช้เมื่อยังไม่มี admin ในระบบ)"""
    print("=== สร้าง Super Admin แรก ===")
    print("⚠️  ฟังก์ชันนี้ใช้เมื่อยังไม่มี admin ในระบบ")
    print()
    
    # ตรวจสอบว่ามี admin อยู่แล้วหรือไม่
    existing_admins = await user_service.get_all_admins()
    if existing_admins:
        print(f"❌ มี Admin อยู่แล้ว {len(existing_admins)} คน")
        print("ใช้โหมดปกติแทน")
        return False
    
    username = input("Super Admin Username: ").strip() or "superadmin"
    password = getpass("Super Admin Password: ").strip()
    if not password:
        password = "SuperAdmin123!"  # default password
        print(f"ใช้ password เริ่มต้น: {password}")
    
    email = input("Super Admin Email: ").strip() or "superadmin@mfu.ac.th"
    
    try:
        from lib.mongodb import get_database
        from utils.security import get_password_hash
        from datetime import datetime
        from bson import ObjectId
        
        db = get_database()
        
        # สร้าง Super Admin โดยตรง
        super_admin_doc = {
            "_id": ObjectId(),
            "nameID": username,
            "username": username,
            "password": get_password_hash(password),
            "email": email,
            "firstName": "Super",
            "lastName": "Admin",
            "department": "IT",
            "role": UserRole.SUPER_ADMIN,
            "groups": ['SuperAdmin', 'Admin'],
            "created": datetime.utcnow(),
            "updated": datetime.utcnow()
        }
        
        await db.get_collection("users").insert_one(super_admin_doc)
        
        print("\n✅ สร้าง Super Admin สำเร็จ!")
        print(f"Username: {username}")
        print(f"Password: {password}")
        print(f"Email: {email}")
        print(f"Role: {UserRole.SUPER_ADMIN}")
        print()
        print("⚠️  กรุณาเปลี่ยน password หลังจาก login ครั้งแรก")
        
        return True
        
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาด: {str(e)}")
        return False

async def list_admins():
    """แสดงรายชื่อ admin ทั้งหมด"""
    print("=== รายชื่อ Admin ทั้งหมด ===")
    
    try:
        admins = await user_service.get_all_admins()
        
        if not admins:
            print("❌ ไม่มี Admin ในระบบ")
            return
        
        print(f"พบ Admin จำนวน {len(admins)} คน:")
        print()
        print(f"{'ID':<25} {'Username':<20} {'Name':<30} {'Department':<20} {'Role':<15}")
        print("-" * 110)
        
        for admin in admins:
            full_name = f"{admin.firstName or ''} {admin.lastName or ''}".strip()
            print(f"{admin.id:<25} {admin.username:<20} {full_name:<30} {admin.department or 'N/A':<20} {admin.role:<15}")
        
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาด: {str(e)}")

async def main():
    """Main function"""
    try:
        # เชื่อมต่อฐานข้อมูล
        await connect_to_mongo()
        print("✅ เชื่อมต่อฐานข้อมูลสำเร็จ")
        print()
        
        # เมนูหลัก
        while True:
            print("=== MFU Learn AI - Admin Management ===")
            print("1. สร้าง Admin ใหม่")
            print("2. แสดงรายชื่อ Admin")
            print("3. สร้าง Super Admin (ใช้เมื่อไม่มี admin)")
            print("4. ออกจากโปรแกรม")
            print()
            
            choice = input("เลือกเมนู (1-4): ").strip()
            
            if choice == "1":
                await create_admin_interactive()
            elif choice == "2":
                await list_admins()
            elif choice == "3":
                await create_super_admin()
            elif choice == "4":
                print("👋 ออกจากโปรแกรม")
                break
            else:
                print("❌ กรุณาเลือก 1-4")
            
            print("\n" + "="*50 + "\n")
        
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาดในการเชื่อมต่อฐานข้อมูล: {str(e)}")
        return False
    
    finally:
        # ปิดการเชื่อมต่อฐานข้อมูล
        await close_mongo_connection()
        print("✅ ปิดการเชื่อมต่อฐานข้อมูลแล้ว")

if __name__ == "__main__":
    # ตรวจสอบว่าอยู่ใน project directory
    if not os.path.exists("backend"):
        print("❌ กรุณารันสคริปต์ในโฟลเดอร์ root ของโปรเจค")
        sys.exit(1)
    
    # รัน script
    asyncio.run(main()) 

File: create_admin.js
#!/usr/bin/env node

/**
 * Create Admin/Super Admin Script
 * Usage: node create_admin.js
 *
 * Environment Variables:
 * - MONGODB_URI: MongoDB connection string
 * - ADMIN_USERNAME: Default admin username
 * - ADMIN_PASSWORD: Default admin password
 * - ADMIN_EMAIL: Default admin email
 * - ADMIN_ROLE: Default role (Admin or SuperAdmin)
 * - ADMIN_DEPARTMENT: Default department
 */

const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const readline = require('readline');
const fs = require('fs');
const path = require('path');

// Load .env file if it exists
function loadEnvFile() {
  const envPath = path.join(__dirname, '.env');
  if (fs.existsSync(envPath)) {
    console.log('📄 Loading environment variables from .env file...');
    const envContent = fs.readFileSync(envPath, 'utf8');
    const envVars = envContent.split('\n').filter(line => {
      return line.trim() && !line.trim().startsWith('#') && line.includes('=');
    });

    envVars.forEach(line => {
      const [key, ...valueParts] = line.split('=');
      const value = valueParts.join('=').replace(/^"|"$/g, '').trim();
      if (key && value && !process.env[key]) {
        process.env[key] = value;
      }
    });
    console.log('✅ Environment variables loaded from .env\n');
  }
}

// User Role enum
const UserRole = {
  ADMIN: 'Admin',
  STAFFS: 'Staffs',
  STUDENTS: 'Students',
  SUPER_ADMIN: 'SuperAdmin'
};

// User Schema (matching the existing schema)
const UserSchema = new mongoose.Schema({
  nameID: { type: String, required: true },
  username: { type: String, required: true, unique: true },
  password: { type: String },
  email: { type: String, required: true },
  firstName: { type: String },
  lastName: { type: String },
  department: { type: String },
  role: {
    type: String,
    enum: Object.values(UserRole),
    default: UserRole.STUDENTS
  },
  groups: { type: [String], default: [] },
  tokenQuota: { type: Number, default: 10000 },
  dailyTokenLimit: { type: Number, default: 10000 },
  created: { type: Date, default: Date.now },
  updated: { type: Date, default: Date.now }
});

const User = mongoose.model('User', UserSchema);

// Create readline interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Helper function to ask questions
function askQuestion(question) {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer.trim());
    });
  });
}

// Helper function to hide password input
function askPassword(question) {
  return new Promise((resolve) => {
    process.stdout.write(question);
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.setEncoding('utf8');

    let password = '';
    process.stdin.on('data', (key) => {
      if (key === '\u0003') { // Ctrl+C
        process.exit();
      }
      if (key === '\r' || key === '\n') { // Enter
        process.stdin.setRawMode(false);
        process.stdin.pause();
        process.stdout.write('\n');
        resolve(password);
      } else if (key === '\u007f') { // Backspace
        if (password.length > 0) {
          password = password.slice(0, -1);
          process.stdout.write('\b \b');
        }
      } else {
        password += key;
        process.stdout.write('*');
      }
    });
  });
}

// Hash password function
async function hashPassword(password) {
  const saltRounds = 12;
  return await bcrypt.hash(password, saltRounds);
}

// Validate email format
function isValidEmail(email) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

// Main function
async function createAdmin() {
  try {
    // Load .env file first
    loadEnvFile();

    console.log('🚀 MFU Learn AI - Admin Creation Script');
    console.log('=====================================\n');

    // Get MongoDB connection string
    // Try different MongoDB environment variable names
    const mongoUri = process.env.MONGODB_URI ||
                    process.env.DATABASE_URL ||
                    await askQuestion('📍 MongoDB URI (default: mongodb://root:1234@localhost:27017/mfu_chatbot?authSource=admin): ') ||
                    'mongodb://root:1234@localhost:27017/mfu_chatbot?authSource=admin';

    console.log('\n🔌 Connecting to MongoDB...');
    await mongoose.connect(mongoUri);
    console.log('✅ Connected to MongoDB successfully!\n');

    // Get user input
    const username = process.env.ADMIN_USERNAME ||
                    await askQuestion('👤 Username: ');

    if (!username) {
      throw new Error('Username is required');
    }

    // Check if user already exists
    const existingUser = await User.findOne({ username });
    if (existingUser) {
      console.log(`❌ User with username "${username}" already exists!`);

      const overwrite = await askQuestion('🔄 Do you want to update this user? (y/N): ');
      if (overwrite.toLowerCase() !== 'y' && overwrite.toLowerCase() !== 'yes') {
        console.log('❌ Operation cancelled.');
        process.exit(1);
      }
    }

    const email = process.env.ADMIN_EMAIL ||
                 await askQuestion('📧 Email: ');

    if (!email) {
      throw new Error('Email is required');
    }

    if (!isValidEmail(email)) {
      throw new Error('Please enter a valid email address');
    }

    const password = process.env.ADMIN_PASSWORD ||
                    await askPassword('🔒 Password: ');

    if (!password) {
      throw new Error('Password is required');
    }

    if (password.length < 6) {
      throw new Error('Password must be at least 6 characters long');
    }

    const firstName = process.env.ADMIN_FIRSTNAME ||
                     await askQuestion('👤 First Name (optional): ') || '';

    const lastName = process.env.ADMIN_LASTNAME ||
                    await askQuestion('👤 Last Name (optional): ') || '';

    const department = process.env.ADMIN_DEPARTMENT ||
                      await askQuestion('🏢 Department (optional): ') || '';

    console.log('\n📋 Available Roles:');
    console.log('1. Admin - Department management permissions');
    console.log('2. SuperAdmin - System-wide management permissions');

    const roleChoice = process.env.ADMIN_ROLE ||
                      await askQuestion('🎯 Choose role (1=Admin, 2=SuperAdmin, default=1): ') || '1';

    let role;
    switch (roleChoice) {
      case '1':
      case 'Admin':
      case 'admin':
        role = UserRole.ADMIN;
        break;
      case '2':
      case 'SuperAdmin':
      case 'superadmin':
        role = UserRole.SUPER_ADMIN;
        break;
      default:
        role = UserRole.ADMIN;
    }

    console.log('\n🔐 Hashing password...');
    const hashedPassword = await hashPassword(password);

    const userData = {
      nameID: username, // Use username as nameID
      username,
      password: hashedPassword,
      email,
      firstName,
      lastName,
      department,
      role,
      groups: [],
      tokenQuota: 100000, // Higher quota for admins
      dailyTokenLimit: 50000, // Higher daily limit for admins
      created: new Date(),
      updated: new Date()
    };

    console.log('\n💾 Creating/Updating admin user...');

    if (existingUser) {
      await User.findOneAndUpdate({ username }, userData, { new: true });
      console.log('✅ Admin user updated successfully!');
    } else {
      const newUser = new User(userData);
      await newUser.save();
      console.log('✅ Admin user created successfully!');
    }

    console.log('\n📋 User Details:');
    console.log(`   Username: ${username}`);
    console.log(`   Email: ${email}`);
    console.log(`   Role: ${role}`);
    console.log(`   Department: ${department || 'Not specified'}`);
    console.log(`   Token Quota: ${userData.tokenQuota}`);
    console.log(`   Daily Limit: ${userData.dailyTokenLimit}`);

    console.log('\n🎉 Admin creation completed successfully!');
    console.log('\n💡 You can now login at: /admin/login');

  } catch (error) {
    console.error('\n❌ Error creating admin:', error.message);
    process.exit(1);
  } finally {
    rl.close();
    if (mongoose.connection.readyState === 1) {
      await mongoose.connection.close();
      console.log('🔌 Database connection closed.');
    }
    process.exit(0);
  }
}

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\n\n👋 Goodbye!');
  rl.close();
  if (mongoose.connection.readyState === 1) {
    mongoose.connection.close();
  }
  process.exit(0);
});

// Run the script
if (require.main === module) {
  createAdmin();
}

module.exports = { createAdmin };